extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,192 @@
+package io.kubernetes.client.extended.controller;
+
+import io.kubernetes.client.extended.controller.reconciler.Reconciler;
+import io.kubernetes.client.extended.wait.Wait;
+import io.kubernetes.client.extended.workqueue.RateLimitingQueue;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultController implements Controller {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultController.class);
+
+  private Reconciler reconciler;
+  private String name;
+  private int workerCount;
+  private ExecutorService workerThreadPool;
+
+  private final RateLimitingQueue<Reconciler.Request> workQueue;
+
+  private final Supplier<Boolean>[] readyFuncs;
+  private Duration readyTimeout = Duration.ofSeconds(30);
+  private Duration readyCheckInternal = Duration.ofSeconds(30);
+
+  public DefaultController(
+      Reconciler reconciler,
+      RateLimitingQueue<Reconciler.Request> workQueue,
+      Supplier<Boolean>... readyFuncs) {
+    this.reconciler = reconciler;
+    this.workQueue = workQueue;
+    this.readyFuncs = readyFuncs;
+  }
+
+  @Override
+  public void run() {
+    if (workerCount <= 0) {
+      log.error("Fail to start controller {}: worker count must be positive.");
+    }
+    if (workerThreadPool == null) {
+      log.error("Fail to start controller {}: missing worker thread-pool.");
+    }
+
+    boolean synced = true;
+    if (this.readyFuncs.length > 0) {
+      synced =
+          Wait.poll(
+              readyCheckInternal,
+              readyTimeout,
+              () -> {
+                boolean ready = true;
+                for (Supplier<Boolean> cacheHasSyncedFunc : this.readyFuncs) {
+                  ready = ready && cacheHasSyncedFunc.get();
+                }
+                return ready;
+              });
+    }
+
+    if (!synced) {
+      log.error("Timed out waiting for cache to be synced.");
+      return;
+    }
+
+    CountDownLatch latch = new CountDownLatch(workerCount);
+    for (int i = 0; i < this.workerCount; i++) {
+      final int workerIndex = i;
+      workerThreadPool.submit(
+          () -> {
+            if (log.isDebugEnabled()) {
```

Comment:
Why do you have an if check here?

Reply:
(and below)

Reply:
it's for improving performance, so that we won't reach the debugging branches on non-debugging deployment.. it's conventional but makes the code less readable on the flip side of the coin

Reply:
No, but isn't there a 100% equivalent check in the `log.debug(...)`

Like I would guess that the first line of code in `log.debug(...)` is:

`if (!log.isDebugEnabled()) return;`

Reply:
> No, but isn't there a 100% equivalent check in the log.debug(...)

that depends on the underlying logging library. slf4j defines a unified interface for logger and other libraries like log4j actually works under it.

Reply:
No, but log.debug(...) shouldn't print if that log level isn't set, so regardless of the implementaton, it has to do the check, right?

Reply:
after roughly googling this, i find it's actually debatable for quite a period of time. some folks insist branching by `log.isDebugEnabled()` is neglectable.. how do you think we consider solve it in a follow-up?

Reply:
I'd really love to remove the code, I think it makes it much harder to read with minimal benefit in performance.

Reply:
removed all the debug condition check